content/posts/metamask-genezio-login.md

```diff
@@ -0,0 +1,218 @@
+If the user is logged in, we retrieve the user’s details. Otherwise, the function throws an error, prompting us to display the ‘Login with Metamask’ button.
+
+```typescript
+{{< filePath >}}client/src/App.tsx{{< /filePath >}}
+useEffect(() => {
+    AuthService.getInstance().userInfo().then((user) => {
+        if (user.address) {
+            getBalance(user.address)
+        } else {
+            throw new Error("No address")
+        }
+    }).catch((e: unknown) => {
+        console.error("Not authenticated. Redirecting to login screen...", e)
+        setData({
+            address: null,
+            balance: null,
+        })
+    })
+}, [getBalance])
+```
+
+Let's see how we implement what happens when the user clicks the 'Login with Metamask' button.
+This action initiates the flow we discussed at the beginning of the article.
+A prerequisite step involves checking whether the user has the Metamask extension installed.
+If the extension is present, it adds the 'ethereum' object to the window property, which we then use to communicate with the Metamask extension.
+
+The first step is to request the wallet's public address using the `eth_requestAccounts` method.
+Once we have it, we send it to the `AuthService` and request a nonce in return.
+Next, we ask the user to sign the nonce to validate that they indeed have access to this public address, using the `personal_sign` method.
+After sending the signature to the `AuthService`, if it is verified as correct, the user will be logged in. The user will remain logged into the web app until they choose to log out.
+
+```typescript
+{{< filePath >}}client/src/App.tsx{{< /filePath >}}
+const loginWithMetamask = async () => {
+    // Check if Metamask is installed
+    if (window.ethereum) {
+        // Fetch the accounts
+        const addresses = await window.ethereum.request({ method: "eth_requestAccounts" })
+        const address = addresses[0]
+
+        // Retrieve a nonce for this address
+        const nonce = await AuthService.getInstance().web3GetNonce(address)
+
+        // Sign the nonce
+        const signature = await window.ethereum.request({
+            method: 'personal_sign',
+            params: [nonce, address]
+        })
+
+        // Login with the signature
+        await AuthService.getInstance().web3Login(address, signature)
+        await getBalance(address)
+    } else {
+        alert("Install Metamask extension!");
+	}
+}
+```
+
+Our app also has a small backend with one function which can be invoked only by authenticated users. If the method is invoked by an authenticated users, it will throw an error and we redirect the user to the login screen.
+
+```typescript
+{{< filePath >}}client/src/App.tsx{{< /filePath >}}
+BackendService.hello("Friend").then((res) => {
+    setSecuredInfo(res)
+}).catch((e) => {
+    console.error(e)
+    logout() 
+})
+```
```

Comment:
It's not really clear what do you want to illustrate with this code snippet.

Maybe add a comment:
```
// Calling the backend method. Only authenticated users can get the sensitive information.
```